### PR TITLE
Interop

### PR DESCRIPTION
**Pull request by giorgiozoppi**
- Add interoperability infrastructure. The idea is to trigger the class serialization to disk from Java in order to read back from C++. This will allow you to use catch to verify each pieces that we are writing without passing through JNI. This mechanism has been integrated in CMake.

- I have started the initial version of the base classes for encoding, the idea here is to use static polymorphism to speed up encoding/decoding. See: https://www.modernescpp.com/index.php/c-is-still-lazy
- I had to use reflection in Java because the original code is doesn't respect interface segregation principle (the method `serializeTo` should be to another interface). So i wanted to minimize code and since it is not strategic code.

### CODE REVIEW

File: include/tsfile/encoding/decoder.h [hunk 1]
```diff
@@ -6,40 +24,54 @@
...
-                iotdb::file::metadata::ts_encoding _type;
-                iotdb::tsfile::encoding::endian_type _endianess;
+                file::ts_encoding _encoding;
+                endian_type _endian;
             };
+
+            template<typename Kind>
+            abstract_decoder<Kind>::abstract_decoder(const file::ts_encoding &type, const endian_type _endian): _encoding(type) {
+
+            }
```
Comment (roberti42) [inline include/tsfile/encoding/decoder.h:49]: What if we keep together the implementation and declaration in this early stage of the project?

File: include/tsfile/read/sequence_reader.h [hunk 2]
```diff
@@ -109,7 +109,7 @@ class sequence_reader {
     encoding::endian_type endian_type() const {
         return _endian_type;
     }
-
+// FIXME: is here shared really needed?
```
Comment (roberti42) [inline include/tsfile/read/sequence_reader.h:112]: maybe not

File: test/build_java.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+GRADLE_BIN=`which gradle`
+JAVA_BIN=`which java`
+echo "Building java interoperability tests"
+cd ${CURRENT_DIR}/iotdb-interop
+${GRADLE_BIN} build
+cd ${CURRENT_DIR}/iotdb-interop/build/libs
+echo "Creating java serialized data"
+java -jar iotdb-interop.jar ${CURRENT_DIR}/iotdb-interop/data
+cd ${CURRENT_DIR}
+
```
Comment (roberti42) [inline test/build_java.sh:12]: First run of `make java_test`:
```
../../test/build_java.sh: line 6: cd: iotdb-native/build-clang/test/iotdb-interop: No such file or directory
```

After creation of iotdb-native/build-clang/test/iotdb-interop, second run:
```
../../test/build_java.sh: line 8: cd: iotdb-native/build-clang/test/iotdb-interop/build/libs: No such file or directory
Creating java serialized data
Error: Unable to access jarfile iotdb-interop.jar
Built target java_test
```

Comment (giorgiozoppi) [reply test/build_java.sh:12]: This is a cmake issue, i try to use with clang.

File: include/tsfile/read/sequence_reader.h (hunk 2 above)
Comment (giorgiozoppi) [reply include/tsfile/read/sequence_reader.h:112]: please evaluate.

File: include/tsfile/encoding/decoder.h (hunk 1 above)
Comment (giorgiozoppi) [reply include/tsfile/encoding/decoder.h:49]: it is ok.

Comment (giorgiozoppi) [reply include/tsfile/encoding/decoder.h:49]: when we do templates is just single header.

**Pull request merged**